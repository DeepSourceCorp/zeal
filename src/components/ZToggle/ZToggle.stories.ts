import '../../assets/css/tailwind.css'
import '../../assets/css/typography.css'
import '../../assets/css/layout.css'
import ZToggle from '../ZToggle/ZToggle.vue'

export default {
  title: 'Form Inputs and Controls: Toggle Button',
  component: ZToggle,
  excludeStories: /.*Data$/
}

export const DefaultToggle = () => ({
  components: { ZToggle },
  data() {
    return {
      value: true
    }
  },

  template: `<div class='padded-container'>
            <div class="input-container p-4">
                <z-toggle :value="value"
                @input="() => this.value = !this.value"></z-toggle>
            </div>
        </div>`
})

export const DisabledToggle = () => ({
  components: { ZToggle },
  data() {
    return {
      value: false
    }
  },

  template: `<div class='padded-container'>
            <div class="input-container p-4">
                <z-toggle :value="value" :disabled="true"
                @input="() => this.value = !this.value"></z-toggle>
            </div>
        </div>`
})

export const DisabledTogglePreSelected = () => ({
  components: { ZToggle },
  data() {
    return {
      value: true
    }
  },

  template: `<div class='padded-container'>
            <div class="input-container p-4">
                <z-toggle :value="value" :disabled="true"
                @input="() => this.value = !this.value"></z-toggle>
            </div>
        </div>`
})
