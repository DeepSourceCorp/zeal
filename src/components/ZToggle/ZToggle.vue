<template>
  <div
    class="inline-block cursor-pointer w-8 h-5 rounded-full focus:outline-none"
    role="checkbox"
    :aria-checked="value.toString()"
    tabindex="0"
    @click="toggle(!value)"
    @keydown.space.prevent="toggle"
    :class="{
      'cursor-not-allowed': this.disabled
    }"
  >
    <!-- prettier-ignore -->
    <span
      class="inline-block rounded-full h-full w-full transition-DEFAULT duration-300 ease-in-out p-0.5 flex items-center"
      :class="{
        'bg-juniper': this.value,
        'bg-ink-100': !this.value || this.disabled
      }"
    >
      <!-- prettier-ignore -->
      <span
        class="inline-block w-3.5 h-3.5 transform rounded-full transition-transform ease-in-out duration-300 shadow-grey"
        :class="{
          'bg-vanilla-400': !this.disabled,
          'bg-vanilla-400 opacity-20': this.disabled,
          'translate-x-full': this.value,
          'translate-x-0': !this.value
        }"
      />
    </span>
  </div>
</template>

<script>
export default {
  name: 'ZToggle',
  props: {
    value: {
      type: Boolean,
      required: true
    },
    disabled: {
      type: Boolean
    }
  },
  model: {
    prop: 'value',
    event: 'input'
  },
  methods: {
    toggle(newValue) {
      if (!this.disabled) this.$emit('input', newValue)
    }
  }
}
</script>
