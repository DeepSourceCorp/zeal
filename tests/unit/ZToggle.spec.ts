import ZToggle from '../../src/components/ZToggle'

import { shallowMount } from '@vue/test-utils'

describe('Toggle', () => {
  it('renders a default toggle', () => {
    const wrapper = shallowMount(ZToggle, {
      propsData: {
        value: true
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders a default toggle in the off state', () => {
    const wrapper = shallowMount(ZToggle, {
      propsData: {
        value: false
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders a disabled toggle', () => {
    const wrapper = shallowMount(ZToggle, {
      propsData: {
        value: false,
        disabled: true
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders a disabled toggle with pre-selected value', () => {
    const wrapper = shallowMount(ZToggle, {
      propsData: {
        value: true,
        disabled: true
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
