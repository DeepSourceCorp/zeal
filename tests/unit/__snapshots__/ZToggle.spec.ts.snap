// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle renders a default toggle 1`] = `<div role="checkbox" aria-checked="true" tabindex="0" class="inline-block w-8 h-5 rounded-full cursor-pointer focus:outline-none"><span class="inline-flex rounded-full h-full w-full transition-DEFAULT duration-300 ease-in-out p-0.5 items-center bg-juniper"><span class="inline-block w-3.5 h-3.5 transform rounded-full transition-transform ease-in-out duration-300 shadow-grey bg-vanilla-100 translate-x-full"></span></span></div>`;

exports[`Toggle renders a disabled toggle 1`] = `<div role="checkbox" aria-checked="false" tabindex="0" class="inline-block w-8 h-5 rounded-full cursor-pointer focus:outline-none cursor-not-allowed"><span class="inline-flex rounded-full h-full w-full transition-DEFAULT duration-300 ease-in-out p-0.5 items-center bg-ink-100"><span class="inline-block w-3.5 h-3.5 transform rounded-full transition-transform ease-in-out duration-300 shadow-grey bg-vanilla-400 translate-x-0"></span></span></div>`;

exports[`Toggle renders a disabled toggle with pre-selected value 1`] = `<div role="checkbox" aria-checked="true" tabindex="0" class="inline-block w-8 h-5 rounded-full cursor-pointer focus:outline-none cursor-not-allowed"><span class="inline-flex rounded-full h-full w-full transition-DEFAULT duration-300 ease-in-out p-0.5 items-center bg-juniper bg-opacity-50"><span class="inline-block w-3.5 h-3.5 transform rounded-full transition-transform ease-in-out duration-300 shadow-grey bg-vanilla-400 translate-x-full"></span></span></div>`;
